pinot-core/src/main/java/org/apache/pinot/core/query/selection/SelectionOperatorUtils.java

```diff
@@ -129,16 +139,55 @@ private SelectionOperatorUtils() {
       }
     } else {
       for (ExpressionContext selectExpression : selectExpressions) {
-        if (expressionSet.add(selectExpression)) {
-          expressions.add(selectExpression);
+        ExpressionContext expression = selectExpression;
+        ExpressionContext newExpression = overrideWithFunctionIndex(indexSegment, pushdownExpressionsMap, expression);
+        if (newExpression != null) {
+          expression = newExpression;
+        }
+        if (expressionSet.add(expression)) {
+          expressions.add(expression);
         }
       }
     }
 
     return expressions;
   }
 
-  /**
+  private static ExpressionContext overrideWithFunctionIndex(IndexSegment indexSegment,
+      Map<Integer, String> pushdownExpressions, ExpressionContext expression) {
+    if (expression.getType() != ExpressionContext.Type.FUNCTION) {
+      return null;
+    }
+    String pushdownExpression = pushdownExpressions.get(expression.hashCode());
+    if (pushdownExpression != null && indexSegment.getDataSource(pushdownExpression) != null) {
+      return ExpressionContext.forIdentifier(pushdownExpression);
+    }
+    List<ExpressionContext> arguments = expression.getFunction().getArguments();
+    for (int i = 0; i < arguments.size(); i++) {
+      ExpressionContext argument = arguments.get(i);
+      ExpressionContext newExpression = overrideWithFunctionIndex(indexSegment, pushdownExpressions, argument);
+      if (newExpression != null) {
+        arguments.set(i, newExpression);
+      }
+    }
+    return null;
+  }
+
+  public static Map<Integer, String> getPushdownExpressionsMap(QueryContext queryContext) {
+    Map<String, String> pushdownExpressions = ImmutableMap.of();
+    try {
+      String pushdownExpressionsOption = queryContext.getQueryOptions().get("pushdown.timestamp.index.expression");
+      if (pushdownExpressionsOption != null) {
+        pushdownExpressions = new ObjectMapper().readValue(pushdownExpressionsOption, Map.class);
```

Comment:
Here also, need to avoid constructing ObjectMapper anywhere but one off initialisation code paths.